notebooks/generate_bracket.py

```diff
@@ -0,0 +1,309 @@
+                                               seed_slot_map[node.value]].values[0][3]
+                next_nodes.append(node.left)
+                next_nodes.append(node.right)
+                counter += 2
+            current_index += 1
+            current_id += 1
+        current_nodes = next_nodes
+
+    num_slots = len(seed_slot_map.keys())
+
+    def get_team_id(seedMap):
+        return (seedMap, df[df['seed'] == seed_slot_map[seedMap]]['teamid'].values[0])
+
+    # Solve bracket using predictions
+    for level in list(reversed(bkt.levels)):
+        for ix, node in enumerate(level[0: len(level) // 2]):
+            team1 = get_team_id(level[ix * 2].value)
+            team2 = get_team_id(level[ix * 2 + 1].value)
+            if team2[1] < team1[1]:
+                temp = team1
+                team1 = team2
+                team2 = temp
+            gid = '{season}_{t1}_{t2}'.format(
+                season=year, t1=team1[1], t2=team2[1])
+            if submit[submit[ID] == gid][PRED].values[0] >= 0.5:
+                level[ix * 2].parent.value = team1[0]
+            else:
+                level[ix * 2].parent.value = team2[0]
+
+    # Create data for writing to image
+    slotdata = []
+    for ix, key in enumerate([b for a in bkt.levels for b in a]):
+        xy = slot_coordinates[year][num_slots - ix]
+        try:
+            st = '{seed} {team}'.format(
+                seed=seed_slot_map[key.value],
+                team=df[df['seed'] == seed_slot_map[key.value]][TEAM].values[0]
+            )
+        except IndexError as e:
+            st = str(seed_slot_map[key.value])
+        slotdata.append((xy, st))
+
+    # Create bracket image
+    # relevant:
+    # https://stackoverflow.com/questions/26649716/how-to-show-pil-image-in-ipython-notebook
+    emptyBracketPath = 'empty_bracket/empty.jpg'
+    img = Image.open(emptyBracketPath)
+    draw = ImageDraw.Draw(img)
+    # font = ImageFont.truetype(<font-file>, <font-size>)
+    # draw.text((x, y),"Sample Text",(r,g,b))
+    for slot in slotdata:
+        draw.text(slot[0], str(slot[1]), (0, 0, 0))
+
+    dpi = 72
+    margin = 0.05  # (5% of the width/height of the figure...)
+    xpixels, ypixels = 940, 700
+
+    # Make a figure big enough to accomodate an axis of xpixels by ypixels
+    # as well as the ticklabels, etc...
+    figsize = (1 + margin) * ypixels / dpi, (1 + margin) * xpixels / dpi
+    fig = plt.figure(figsize=figsize, dpi=dpi)
+    # Make the axis the right size...
+    ax = fig.add_axes([margin, margin, 1 - 2*margin, 1 - 2*margin])
+
+    ax.imshow(np.asarray(img))
+    # plt.show() # for in notebook
+    img.save(output_path)
+
+# Sample call
+
+# b = build_bracket(
+#     output_path="2021_bracket.png",
+#     teamsPath="../../data/MTeams.csv",
+#     seedsPath="../../data/MNCAATourneySeeds.csv",
+#     slotsPath="../../data/MNCAATourneySlots.csv",
+#     submissionPath="ncaa-march-madness-submission.csv",
+#     year=2021
+# )
```

Comment:
The bracketed stuff was expecting slightly different naming conventions for some columns, I don't think it's been updated. I just kinda borrowed the whole thing, made some edits, and confirmed it was still working.